### ISSUE TITLE
[RFC] Standardized type conversion (From/Into)

### ISSUE THREAD

**Issue by z64** [crystal-lang/crystal#8511]
I think it would be great if std could provide an answer to the question(s):

- "I want type Foo that can be converted into instances of Bar"

or

- "I want type Bar that can interpret instances of Bar from Foo"

The general issue is that I believe the language - compiler, std, or documentation - does not advise how to best solve these problems.

Having reviewed shards and answered a lot of questions about Crystal over the years, the lack of guidelines here has caused a bit of a mess, because Crystal gives you a lot of options to do this, and grasping which one is best isn't inferred well, according to my surveys. I feel it has been a sort of "silent issue" as everyone more or less just does their own thing.

# Ecosystem Examples

Here's the most common methods that any user might pick - particularly beginners - along with some observations. Each of which may be a pro or con depending on who you ask. This is certainly not exhaustive.

Considering `Foo` and `Bar`, which may be fully or only partially compatible field-wise, and the conversion is well defined:

## 1. `T.new(other : U) : T`

```cr
class Foo
  def self.new(bar : Bar) : Foo
    a = bar.a
    b = bar.b
    # Derived or default value that bar does not provide
    c = "some string"
    Foo.new(a, b, c)
  end
end
```

- `Foo.new` is overloaded, WYSIWYG (contrast with (2) below).
- Private & protected access to `Foo`'s members at the class level.
- Public access to `Bar`'s members.
- Does not necessarily have to be called `new`.
- Reuses the "default" constructor likely already defined in terms of itself, `#initialize(@a, @b, @c)`

## 2. `T#initialize(other : U)`

```cr
class Foo
  def initialize(bar : Bar)
    @a = bar.a
    @b = bar.b
    # Derived or default value that bar does not provide
    @c = "some string"
  end
end
```

- Compiler's `initialize` magic comes into play: It defines a `self.new` method for you, all ivars must be initialized, etc. 
- A lot of users taking this approach will habitually call *another* `initialize` method from within this one if setting up the subject type is complicated, or just a lot of lines of code, in the example above they might replace the body with: 

```cr
def initialize(bar : Bar)
  initialize(bar.a, bar.b, "some string")
end
```

- Sometimes the above leads to a "tree" of `initialize` methods called.
- Unrestricted access to `self`'s members
- Public access to `Bar`'s members

## 3. `T#to_u` / `U#to_t`

```cr
class Foo
  def to_bar : Bar
    Bar.new(@a, @b) 
  end
end

class Bar
  def to_foo : Foo
    Foo.new(@a, @b, "some string") 
  end
end
```

- Reserves `#to_bar` as a method name, since it takes no arguments.
- Some users will use different naming convetions: `#as_bar`, `#convert_to_bar`, or just `#bar`
- Unrestricted access to `self`'s members

## Examples Summary

There are others I've seen, but these are probably the 3 that most people use. My opinions on these approaches:

- (1) I don't see too often. It seems to take a while for users to pick up that they can overload `.new`,  or they often do not end up preferring it over (2). In my opinion, this is the best out of the three. The visibility is acceptable in most if not all cases, and you get good compiler errors if the receiving `initialize` changes during refactoring. It makes more sense as a static method to me, and infers the existence of a basic provide-each-field constructor.

- (2) Is probably the most common, but very prone to "trees of `initialize`" for complicated types as I mentioned. This is likely most common because that is how new users are taught to write constructors, so they don't treat a conversion like this any differently. This often leads to hard to follow code in constructors, and users running into compiler errors because they forgot to set an ivar - either when writing or in refactoring the type later, sometimes leading to inconsistent solutions. They might set the missing var, defer to another `initialize`, or do something like add a default argument.

- (3) Slightly less common than (2). Often these users saw methods like `String#to_i` and followed suit for `#to_xyz`, or `JSON::Any`'s `#as_xyz` methods. Compiler errors in refactoring should be comparable to (1). I personally don't like this over (1) as every method has a different name, and it's up to the author to be consistent, let alone between multiple authors.

- Aside: I also see a lot of users trying things like `String.as(Int32)` before finding `String#to_i`. This is of course a different issue entirely, and a small roadbump in learning Crystal, but I think it shows potential value in the proposed solutions below.

# Solutions

## A. Generic API

My general proposal is that a generic API (as in employing generics) for type conversion should be provided in order to:

1. **Encourage a consistent convention across the ecosystem**
2. **Allow conversion to be reasoned about**

In short, allowing any library author to provide something like the following:

```cr
# Does something.
# Accepts any type that can (perhaps lazily) be converted into `Bar`.
def do_something(subject : Into(Bar))
   bar = subject.into(Bar)
   # Do things with bar
end
```

wherein the library user would then implement `Into(Bar)` for their type.

And additionally `From` to apply conversion the other way around. This is a little more restrictive because of Crystal's generics / inheritance model, it would have to rely on `Into`:

```cr
module From
  def from(other : Into(self))
    other.into(self)
  end
end

module Into(T)
  abstract def into(klass : T.class) : T
end

class Bar; end
  
class Foo
  extend From
  include Into(Bar)

  def initialize(@string : String)
  end
  
  def into(bar : Bar.class) : Bar
    Bar.new(@string.to_i)
  end
end

class Bar
  extend From
  include Into(Foo)

  def initialize(@int : Int32)
  end
  
  def into(foo : Foo.class) : Foo
    Foo.new(@int.to_s)
  end
end

bar = Bar.new(5)
p(Foo.from(bar))
p(bar.into(Foo))

foo = Foo.new("6")
p(Bar.from(foo))
p(foo.into(Bar))
```

`From(T)` would be great, but I don't think this can be expressed reasonably on the metaclass. Please feel free to correct me. I also acknowledge the general attitude towards `extend`, and perhaps `from` could be provided by other means (or not at all).

This concept is borrowed from and has been largely successful in Rust: 
- https://doc.rust-lang.org/stable/std/convert/trait.From.html
- https://doc.rust-lang.org/stable/std/convert/trait.Into.html

There are uncountable Rust libraries that reason about `From`/`Into` in their interfaces, and it has wholly avoided this situation of "too many solutions" in Crystal.

## Common methods

We can keep common conversion methods as delegations to these implementations:

```cr
class String
  include Into(Int32)

  def into(klass : Int32.class) : Int32
    # ..
  end

  def to_i
    into(Int32)
  end
end
```

## (De)serialization

This may also help provide a more sustainable API for things like the existing `Object.from_json` / `Object#to_json`. Looking in from outside of Crystal, it is awkward having such a "specific" serialization written into every type, which per my examples above, does not set a great standard for others in the ecosystem. ("Does everyone need write `#to_my_format` on every type?")

I acknowledge how ubiquitous JSON is for many of Crystal's users, but it raises other aforementioned questions.

A similar general `Serialize` / `Deserialize` contract would be great to have, but that's tangent to this issue's scope - `From`/`Into(T)` could be a suitable, but maybe limiting entrypoint for some use cases:

```cr
foo.into(JSONString) # "{ .. }"
```

## B. Improve documentation

It would otherwise go a long way to talk about these concepts in the language reference. Even if we don't conform to something in std, a "best practice" picking from something like the examples 1/2/3 above, discussing their pros/cons, or other guideline would help smooth the disparity across shards and help beginners.

I think the main pattern that should be advised against is multiple (nested) calls through `initialize` methods, as this has done the most harm in code that I've read in terms of readability and maintenance burden.

---

That's about it - thanks for reading. 

I recognize the current conversion idioms stem from Ruby. As someone who came from Ruby, to Crystal, and then to Rust - coming back to Crystal for some projects has left consistency like this to be desired where it would be possible for std types to do more.

**Comment by straight-shoota** [comment]
I think one of the problems is communicating the difference between overloading `#initialize` vs. overloading `.new`. This should probably be discussed in learning materials. So a big YES to solution B.
Essentially, most overloads should be for `.new`, especially for the use cases discussed here. However, there might also be rare cases where initializing an instance from a different type requires some customized behaviour that must be implemented in a `#initialize`. So in general (1) and (2) could be seen as the same thing and (1) should be most commonly used.

At least in theory, (3) should be pretty straightforward actually. The regular naming scheme is `to_#{type}`. `to` signals a conversion and is followed by the target of the conversion. Optional qualifiers can be used for clarification.
Such a method should explicitly not be called `as`, because `as` is used to describe a cast (i.e. changing the interpretation of a value) whereas `to` means a conversion (i.e. changing the value). `#to_bool` returns the value converted to `Bool`, whereas `#as_bool` returns the value interpreted as `Bool`.

IMO both methods are suitable and there's not clear benefit to either of them. I suppose it mostly depends on which side offers the conversion logic, the source or the sink. This is especially relevant when both are in different code bases.

I actually don't see a direct connection between your problem description and solution A. The examples are all about explicit methods for converting between types, but solution A is about a mechanism for implicit conversions. Let's not confuse that. In some cases, explicit conversions might also work implicitly and maybe are only explicit because there's no implicit one. But more often than not, implicit conversions are not a solution. Converting between types usually creates some kind of friction. If the types were equivalent, there wouldn't be a reason to have distinct types.
So I'm not quite sure how useful a generic mechanism for implicit conversions would actually be.

I'd be interested in real world examples for Rust's `From` and `Into`, though. The example for From is about error handling, which is already pretty different in Crystal.

**Comment by oprypin** [comment]
I would not agree about a general API to convert between types, it's very easy to lose track of semantics then. Surely there won't always be exactly one way to interpret one type as the other?

**Comment by z64** [comment]
@straight-shoota Yep; fully agree on your first two paragraphs. The only way I've come to that same conclusion is just by reading lots and lots of code and feeling the pitfalls of the other methods :smile_cat: 
So yeah, an addendum to the language reference I think would be a great minimum action here.

> I actually don't see a direct connection between your problem description and solution A. 

The connection is just that it bakes a solution for performing conversions into the standard library instead of by having to teach a recommended method and hope that users follow it, as well as being able to reason about conversions. 

It takes any question out of something like `foo.to_s` - what is "s"? and in turn makes it an explicit `foo.into(String)`

You might have something like a socket wrapper for an API that lets you send some raw data:
```cr
def send(data : Into(Bytes))
  # ..
end
```
Thus clearing the callsite:
```cr
socket.send(my_object)
```
and not relying on duck typing something like `#to_slice`, since you can only document something like that in a comment. 

Even if the compiler tells you "`Error: undefined method 'to_slice' for Foo`", that doesn't give you enough feedback to know what `to_slice` needs to return; you have to read the source or documentation. Restricting on `Into(T)` tells you exactly what you need to implement, and there is no debate about how to implement it, what to name the method, etc.

I'm out of time at the moment so I can't dig out some good rust examples. Will try to check back in later.

**Comment by ysbaddaden** [comment]
> Reuses the "default" constructor likely already defined in terms of itself, #initialize

Please note that `#initialize` isnt a constructor but an initializer; it initializes previously allocated memory. `.new` is the constructor: it allocates then calls initialize (among other things). There should be as many constructors as needed, but only one (or a few) initializers.

Maybe I don't understand the problem/proposal, but from/into sounds like unrequited complexity. Its moving construction of a type A to another type B (into), but eventually brings it back with a delegation to B constructing A (from -> into) which is working backwards, or is far too formal for Crystal.

Each type should construct/initialize itself, but it all depends on the API logic. Crystal, like Ruby, is very flexible but follows conventional rules:
- Do we create A from B? Then we expect A to handle that logic in a constructor (`.new`, `.from` or preferably an explicit name), for example `Debug::ELF.read(io)`. Having `IO#into(Debug::ELF)` would pollute IO, in addition to moving the constructor logic where it doesn't belong.
- Do we transform B into A? Then B is responsible to build A (and A doesn't care about B) and we expect B to have a `#to_a` method that will construct an A, for example `Time#to_unix`. Having `Int64.from(Time)` makes no sense (what would we expect the i64 number to be?).
- Do we interpret B as being an A? Then similarly we expect B to have a `#as_a` method that casts B to A, for example `JSON::Type#as_s`. Again, having a `String.from(JSON::Type)` would be *very* confusing (transform? inspect?), and `JSON::Type.into(String)` would imply a transformation when there is none.

Now, from/into may have some usages, but it's impossible to understand them with generic Foo Bar classes and no concrete use cases. I don't see the benefit. The String to Int32 is hardly convincing, dince there is already the much simpler `#to_i` idiom and adding `#into(Int32)` would introduce a duplication...

**Comment by watzon** [comment]
I believe the purpose of this was to try and find a Crystal solution to the interface problem. Rust solves this problem with Traits, but in Crystal all we have is modules. Modules aren't a terrible solution, but take this example:

```crystal
require "json"

module Jsonable
  abstract def to_json(json : JSON::Builder)
end

class FooBar
  getter foobar : Array(Jsonable)
  
  def initialize(@foobar)
  end
end

arr = [] of Jsonable
arr << "Hello"
arr << 3.14
arr << {"1", "2", 3}

fb = FooBar.new(arr)
pp fb
```

This code will not compile because `String`, `Float64`, and `Tuple(*T)` don't include the `Jsonable` module. Even though it's very clear to the user what should happen, without explicitly opening every type that fits the interface there's no way to make this code work.

One solution would be to do exactly what's being proposed here, or some variation of it. I kind of like it, because it seems like a more Crystally solution. Another option would be to create an `Interface` type for generic structures like this. Then you could define interfaces like this:

```crystal
interface Jsonable
  abstract def to_json(json : JSON::Builder) : String
end
```

The benefit I see to this approach is that it doesn't really require breaking changes, or any changes really to the standard library besides adding this itself. It puts the burden of implementation itself on the user, which they can do as needed.

In my humble opinion Crystal does need something like this though, as strict as its type system is. I can think of tons of examples off the top of my head where, in my own code, having interfaces would save me a lot of trouble.

**Comment by Blacksmoke16** [comment]
Wouldn't you still need to implement this interface on every type that fits it?  At which point you could use a module and get the same effect?

**Comment by watzon** [comment]
@Blacksmoke16 that's the thing. The whole idea of the interface type would be so you don't have to implement it everywhere. Instead it should be considered a limited subset of all types that define the requisite methods. So in my `Jsonable` example it would accept any type that has a `#to_json` method matching the signature provided, and that would be the only method available when using it as a type. So:

```crystal
interface Numeric
  abstract def to_i : Int32
end

class Foo
  def initialize(@bar : Numeric)
    # this should work
    puts @bar.to_i

    # this should either not work, or be dependent on the types being passed in.
    puts @bar.to_s
  end
end
```

**Comment by Blacksmoke16** [comment]
Ahh, gotcha.  Essentially like a fancy alias that resolves to types that match it.

**Comment by watzon** [comment]
Essentially yes. This would make some types a lot easier to handle like, as in the above case, the situation where you just want something that resembles an `Int32` and can be cast to one. Saves you having to know every possible type that can be cast to an `Int32` and make an alias for it.

It could get even more complicated as you add additional types though. To the point where an alias would have to be very specific, and most definitely wouldn't cover user defined types.

**Comment by RX14** [comment]
The difference with crystal is you can reopen any class and make it implement a module anywhere in your code. You don't need "permission" like in many other languages, so an explicit `include` is fine.

**Comment by Blacksmoke16** [comment]
Relaying a thought from Gitter:

One thought I had at some point was introducing the concept of an actual `interface`.

```cr
interface StringConverter
  def convert(value : String) : String
end
```

Essentially this would be the same thing as:
```cr
module StringConverter
  abstract def convert(value : String) : String
end
```

but would assume methods to be `abstract`, and prevent you from defining actual methods/types within it.  Only methods/constants.  Usage would be similar to modules but would be like

```cr
class ReverseConverter
  implement StringConverter
end
```

Benefits are I suppose it makes intent more clear, plus preventing someone from adding stuff to your interface.  If it's worth the extra logic in the compiler and such for, I've yet to be convinced.

**Comment by watzon** [comment]
My main point is that there are cases in which you don't want to, or aren't able to, monkey patch all the classes you may want. Also monkey patching is considered an anti-pattern is it not? I've heard it both ways.

**Comment by straight-shoota** [comment]
@Blacksmoke16 I can see litteraly no benefit at all with that approach. It's just another syntax for what already exists and it's only more restrictive, less flexible.

I kind of like the idea of automatically matching interfaces. I'm not sure they make sense even exploring, let alone introducing it to the language. But it may be a useful feature.

It can be understood as an extension to `responds_to` which takes several methods into account. And that fits pretty well with the duck-typing approach. Often it's not sufficient to see if something can quack but you also need some other properties to treat it as a duck.

**Comment by watzon** [comment]
That's exactly my point @straight-shoota, thanks for laying it out more eloquently than I could

**Comment by RX14** [comment]
If we were making crystal from scratch, it'd probably be a good idea.

Unfortunately, it doesn't mean it's a good idea to introduce it unless we remove something else...

**Comment by caspiano** [comment]
Related https://github.com/crystal-lang/crystal/issues/12643

**Comment by HertzDevil** [comment]
`into` there refers to [this](https://hexdocs.pm/elixir/Enum.html#into/3), it has nothing to do with general object conversion.

**Comment by caspiano** [comment]
[My example here](https://github.com/crystal-lang/crystal/issues/12643#issuecomment-1288968063) does not explicitly reference generic object conversion but seemed like it could be generalised with some more thought.

It does look like the `From` trait in rust does cover conversions between collection types, see [this method from an array to `HashSet`](https://doc.rust-lang.org/stable/std/convert/trait.From.html#impl-From%3C%5BT%3B%20N%5D%3E-for-HashSet%3CT%2C%20RandomState%3E)
